mautic/mautic
# Segment filter refactoring

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Code enhancement of existing feature
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | Y

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

The list of segment operators and functionality around that was hard-coded in the LeadBundle and not open for extension by plugins. This PR refactors that and adds new events that the plugins can hook into.

Most of the code changes are refactoring hard-coded switches into events. But there is one change in UI that may be noticeable. Every time you select an operator in a segment filter then the filter field(s) is being regenerated on server. It used to be hard-coded in JS what field type should show up. It was impossible to plugins to modify it. Now any plugin can.

#### Steps to test this PR:
1. Ensure you have multiple segments with multiple filters already before you check out this PR so you could test that existing segments still work.
2. Once you check out this branch, run the `bin/console mautic:segments:update` command and check that the command worked as before and no contacts were added or removed. The contact count for each existing segment should be the same.
3. Edit an existing segment and check that the filter values are as it used to be. You should check selects, multiselects, tags, segment membership, owner, text, boolean and country filters at least.
4. Try to create multiple segments with different filters at least type of mentioned in the step 3. You should have no problem create those filters, save them and the segments should contain the right contacts once you run the update command.
5. Test also dynamic email content filters. Should work as before.
6. Test also dynamic web content filters. Should work as before.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/8649"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
